addons/mail_group/tests/test_mail_group.py

```diff
@@ -29,23 +32,72 @@ def test_constraint_valid_email(self):
         with self.assertRaises(ValidationError, msg="Moderators must have an email"):
             mail_group.moderator_ids |= user_without_email
 
+    def test_find_member(self):
+        member_1 = self.test_group_member_1
+        email = member_1.email_normalized
+
+        partner_2 = self.user_portal.partner_id
+        partner_2.email = '"Bob" <%s>' % email
+        member_2 = self.env['mail.group.member'].create({
+            'partner_id': partner_2.id,
+            'mail_group_id': self.test_group.id,
+        })
+
+        partner_3 = self.partner_root
+        partner_3.email = '"Bob" <%s>' % email
+        member_3 = self.env['mail.group.member'].create({
+            'partner_id': partner_3.id,
+            'mail_group_id': self.test_group.id,
+        })
+
+        member = self.test_group._find_member(email)
+        self.assertEqual(member, member_1, 'When no partner is provided, return the member without partner in priority')
+
+        member = self.test_group._find_member(email, partner_2.id)
+        self.assertEqual(member, member_2, 'Should return the member with the right partner')
+
+        member = self.test_group._find_member(email, partner_3.id)
+        self.assertEqual(member, member_3, 'Should return the member with the right partner')
+
+        member_2.unlink()
+        member = self.test_group._find_member(email, partner_2.id)
+        self.assertEqual(member, member_1, 'Should return the member without partner')
+
+        member_1.unlink()
+        member = self.test_group._find_member(email, partner_2.id)
+        self.assertFalse(member, 'Should not return any member because the only one with the same email has a different partner')
+
     @users('employee')
     def test_join_group(self):
-        mail_group = self.env['mail.group'].browse(self.test_group.ids)
+        mail_group = self.test_group
         self.assertEqual(len(mail_group.member_ids), 4)
 
         mail_group._join_group('"Jack" <jack@test.com>')
         self.assertEqual(len(mail_group.member_ids), 5)
+        self.assertTrue(mail_group._find_member('"Test" <jack@test.com>'))
 
-        with self.assertRaises(UserError):
-            mail_group._join_group('"Jack the developer" <jack@test.com>')
+        mail_group._join_group('"Jack the developer" <jack@test.com>')
         self.assertEqual(len(mail_group.member_ids), 5, 'Should not have added the duplicated email')
 
+        # Join a group with a different email than the partner
```

Comment:
Precisely this test check that the email is forced to the email of the partner to avoid those tricky situation, so even if this method is called with a different email, it will be forced to the email of the partner